### PR TITLE
Add Addrinfo#==

### PR DESCRIPTION
**Pull request by nobu**

### CODE REVIEW

**Comment by ioquatix** [comment]
It’s a good idea!

**Comment by ioquatix** [comment]
If you merge this can you also close https://bugs.ruby-lang.org/issues/13593

File: ext/socket/raddrinfo.c [hunk 1]
```diff
@@ -1707,6 +1715,55 @@ addrinfo_mload(VALUE self, VALUE ary)
...
+        long l2 = unixsocket_len(r2);
+        if (l1 != l2) return Qfalse;
+        if (memcmp(&r1->addr.un.sun_path, &r2->addr.un.sun_path, l1))
+            return Qfalse;
+        break;
+      }
+#endif
+
+      default:
+        if (memcmp(&r1->addr.addr, &r2->addr.addr, r1->sockaddr_len))
```
Comment (shyouhei) [inline ext/socket/raddrinfo.c]: No, this doesn't work.  For instance in Linux, `struct sockaddr_in` is defined like this:
https://github.com/torvalds/linux/blob/master/include/uapi/linux/in.h

Here, `sin_port` (which is 16 bits big endian) is _followed by_ `sin_addr` (which is 32bit big endian).  They don't agree with alignments.  There are uninitialized padding bits in between them.  Comparing them using `memcmp` detects false negative.

Comment (nobu) [reply ext/socket/raddrinfo.c]: Implemented comparisons of INET and INET6.
What about other families? raise an exception?

Comment (shyouhei) [reply ext/socket/raddrinfo.c]: Persuade @akr it you think raising exceptions is appropriate.

File: ext/socket/raddrinfo.c [hunk 2]
```diff
@@ -1712,6 +1712,81 @@ addrinfo_mload(VALUE self, VALUE ary)
...
+
+    if (CLASS_OF(self) != CLASS_OF(other)) return Qfalse;
+    r1 = get_addrinfo(self);
+    r2 = get_addrinfo(other);
+    if (r1->pfamily != r2->pfamily) return Qfalse;
+    if (r1->socktype != r2->socktype) return Qfalse;
+    if (r1->protocol != r2->protocol) return Qfalse;
+    if (r1->sockaddr_len != r2->sockaddr_len) return Qfalse;
+
+    if (!rb_str_equal(r1->inspectname, r2->inspectname)) return Qfalse;
```
Comment (shyouhei) [inline ext/socket/raddrinfo.c]: It seems comparing `inspectname` is a wrong idea.

**Comment by shyouhei** [comment]
So the problem is, the is no solid definition as to when two `struct addrinfo` are considered identical.  C has no such thing like equality of structs.  Nor addrinfo-related RFCs.  In order to implement this we have to employ our own "view" of what are distinct and what are identical.  In doing so a close inspection of the use-case of this feature is needed. i.e. "What's good about it?  Why do we need this feature?"

**Comment by hsbt** [comment]
https://bugs.ruby-lang.org/issues/13593 is already rejected. I'll close this.

**Pull request closed without merging**